SU2_CFD/src/solvers/CEulerSolver.cpp

```diff
@@ -3122,28 +3125,60 @@ void CEulerSolver::Upwind_Residual(CGeometry *geometry, CSolver **solver_contain
 
       for (iVar = 0; iVar < nPrimVarGrad; iVar++) {
 
-        su2double Project_Grad_i = 0.0;
-        su2double Project_Grad_j = 0.0;
+        const su2double V_ij = V_j[iVar] - V_i[iVar];
+
+        su2double Project_Grad_i = 0.5*Kappa*V_ij;
+        su2double Project_Grad_j = 0.5*Kappa*V_ij;
 
         for (iDim = 0; iDim < nDim; iDim++) {
-          Project_Grad_i += Vector_ij[iDim]*Gradient_i[iVar][iDim];
-          Project_Grad_j -= Vector_ij[iDim]*Gradient_j[iVar][iDim];
+          Project_Grad_i += (1.0-Kappa)*Vector_ij[iDim]*Gradient_i[iVar][iDim];
+          Project_Grad_j += (1.0-Kappa)*Vector_ij[iDim]*Gradient_j[iVar][iDim];
         }
 
         if (limiter) {
           if (van_albada) {
-            su2double V_ij = V_j[iVar] - V_i[iVar];
-            Limiter_i[iVar] = V_ij*( 2.0*Project_Grad_i + V_ij) / (4*pow(Project_Grad_i, 2) + pow(V_ij, 2) + EPS);
-            Limiter_j[iVar] = V_ij*(-2.0*Project_Grad_j + V_ij) / (4*pow(Project_Grad_j, 2) + pow(V_ij, 2) + EPS);
+            Project_Grad_i *= V_ij*(2.0*Project_Grad_i + V_ij) / (4*pow(Project_Grad_i, 2) + pow(V_ij, 2) + EPS);
+            Project_Grad_j *= V_ij*(2.0*Project_Grad_j + V_ij) / (4*pow(Project_Grad_j, 2) + pow(V_ij, 2) + EPS);
           }
-          Primitive_i[iVar] = V_i[iVar] + Limiter_i[iVar]*Project_Grad_i;
-          Primitive_j[iVar] = V_j[iVar] + Limiter_j[iVar]*Project_Grad_j;
-        }
-        else {
-          Primitive_i[iVar] = V_i[iVar] + Project_Grad_i;
-          Primitive_j[iVar] = V_j[iVar] + Project_Grad_j;
-        }
+          else if (piperno) {
+            const su2double R_i = 0.5*V_ij/(Project_Grad_i+EPS);
+            const su2double R_j = 0.5*V_ij/(Project_Grad_j+EPS);
+
+            const su2double InvR_i = (Project_Grad_i)/(0.5*V_ij+EPS);
+            const su2double InvR_j = (Project_Grad_j)/(0.5*V_ij+EPS);
```

Comment:
I think adding EPS here will not protect against division by zero, also if InvR is the inverse of R...